### PR TITLE
Remove empty unchained function calls

### PR DESCRIPTION
**Pull request by JulissaDantes**
For the cases where the constructor is empty we dont wqant the chained function declaration, nor the unchained calls, this PR includes:

- Remove unchained function declaration
- Remove call from contract chained declaration if it doesnt have parameters
- Remove from inheritance call if it doesnt have parameters

### CODE REVIEW

File: src/transform.test.ts.md [hunk 1]
```diff
@@ -300,7 +304,49 @@ Generated by [AVA](https://avajs.dev).
...
+    ␊
+        modifier hasModifier(){␊
+            _;␊
+        }␊
+    ␊
+        function __Foo8_init() internal onlyInitializing {␊
+            __Foo8_init_unchained();␊
+        }␊
+    ␊
+        function __Foo8_init_unchained() internal onlyInitializing {␊
```
Comment (frangio) [inline src/transform.test.ts.md:318]: I've realized there is a preexisting bug: the modifier was removed!

**Comment by JulissaDantes** [comment]
Update: Must antain implementation but still remove calls to empty unchained

File: src/index.test.ts.md [hunk 2]
```diff
@@ -388,7 +383,6 @@ Generated by [AVA](https://avajs.dev).
       ␊
       contract StringConstructorUpgradeable is Initializable {␊
           function __StringConstructor_init(string memory message) internal onlyInitializing {␊
-              __StringConstructor_init_unchained(message);␊
```
Comment (frangio) [inline src/index.test.ts.md:391]: This is going to reintroduce unused parameter warnings. Notice how `message` is no longer used in the function body.

File: src/transform.test.ts.md (hunk 1 above)
Comment (Amxx) [reply src/transform.test.ts.md:330]: Note, the modifier might have side effects, so it should NOT be added to both the chained and the unchained versions.
I suggest it goes in the unchained.

File: src/transformations/utils/build-super-calls-for-chain.ts [hunk 3]
```diff
@@ -110,6 +109,32 @@ function isImplicitlyConstructed(contract: ContractDefinition): boolean {
 
   return (
     contract.contractKind === 'contract' &&
-    (ctor == undefined || ctor.parameters.parameters.length === 0)
+    (ctor == undefined ||
+      ctor.parameters.parameters.length === 0 )
+  );
+}
```
Comment (Amxx) [inline src/transformations/utils/build-super-calls-for-chain.ts:115]: This was only modified by the linter (i guess), but it makes me wonder. Should there be a case about the modifier.

`constructor() mySideEffectModifier() {}` is not implicitly constructed.

Comment (Amxx) [reply src/transformations/utils/build-super-calls-for-chain.ts:115]: or maybe in that case the contract will not be `isEmpty`, and that is good enough

File: src/transformations/utils/build-super-calls-for-chain.ts [hunk 4]
```diff
@@ -110,6 +109,32 @@ function isImplicitlyConstructed(contract: ContractDefinition): boolean {
...
+  const ctor = getConstructor(contract);
+  const varInitNodes = [...findAll('VariableDeclaration', contract)].filter(
+    v => v.stateVariable && v.value && !v.constant && !hasOverride(v, 'state-variable-assignment'),
+  );
+  const modifiers = [];
+  if(ctor){
+    // we only include modifiers that don't reference base contracts
+    for (const call of ctor.modifiers) {
+      const { referencedDeclaration } = call.modifierName;
+      if (referencedDeclaration != null && !ctor.linearizedBaseContracts.includes(referencedDeclaration)) {//is not inheritance
```
Comment (Amxx) [inline src/transformations/utils/build-super-calls-for-chain.ts:127]: I don't understand the `!ctor.linearizedBaseContracts.includes(referencedDeclaration))` part.

what does it mean for a modifier to be in the linearizedBaseContracts ?

File: src/transformations/utils/build-super-calls-for-chain.ts [hunk 5]
```diff
@@ -110,6 +109,32 @@ function isImplicitlyConstructed(contract: ContractDefinition): boolean {
 
   return (
     contract.contractKind === 'contract' &&
-    (ctor == undefined || ctor.parameters.parameters.length === 0)
+    (ctor == undefined ||
+      ctor.parameters.parameters.length === 0 )
+  );
+}
+
+function isEmpty(contract: ContractDefinition): boolean {
```
Comment (Amxx) [inline src/transformations/utils/build-super-calls-for-chain.ts:117]: Why not add
`if (contract.contractKind !== 'contract) { return false; }`
at the begining?

That would somehow simplify the return at the end, and give more confidence about the object we are working on.

Comment (Amxx) [reply src/transformations/utils/build-super-calls-for-chain.ts:117]: I think this all function would be clearer if it was written

```
function isEmpty(contract: ContractDefinition): boolean {
  if (contract.contractKind !== 'contract') { return false; }

  const ctor = getConstructor(contract);

  const hasVarInitNodes = [...findAll('VariableDeclaration', contract)].some(
    v => v.stateVariable && v.value && !v.constant && !hasOverride(v, 'state-variable-assignment'),
  );
  const hasModifiers = ctor.modifiers.some(
    call => call.modifierName.referencedDeclaration != null && !ctor.linearizedBaseContracts.includes(call.modifierName.referencedDeclaration)
  );

  return ctor?.body?.statements?.length == undefined && !hasVarInitNodes && !hasModifiers;
}
```

File: src/transformations/utils/build-super-calls-for-chain.ts [hunk 6]
```diff
@@ -110,6 +109,32 @@ function isImplicitlyConstructed(contract: ContractDefinition): boolean {
...
+      const { referencedDeclaration } = call.modifierName;
+      if (referencedDeclaration != null && !ctor.linearizedBaseContracts.includes(referencedDeclaration)) {//is not inheritance
+        modifiers.push({ call });
+      }
+    }
+  }
+
+  return (
+    contract.contractKind === 'contract' &&
+    (ctor?.body?.statements?.length == undefined &&
```
Comment (Amxx) [inline src/transformations/utils/build-super-calls-for-chain.ts:135]: Here we protect against ctor not including a "body" field, but we don't have the same check against ctor not having a "modifiers" field

File: src/transformations/utils/build-super-calls-for-chain.ts (hunk 4 above)
Comment (JulissaDantes) [reply src/transformations/utils/build-super-calls-for-chain.ts]: When the ast filters by modifier it takes both modifiers, as in `modifier hasmodifier(){_;}`, and base contract contructors, so in this case `constructor (string name, string symbol) ERC20(name, symbol) hasmodifer` the `.modifiers` will bring both as modifiers, but for our check we only need to know if those are actual modifiers.

This was better addressed on the last commit.

File: src/transform.test.ts.md (hunk 1 above)
Comment (JulissaDantes) [reply src/transform.test.ts.md:330]: Yes that is what we will do, include the custom modififers only on the unchained method, that will not be included on this PR but in a following one

File: src/transformations/utils/build-super-calls-for-chain.ts (hunk 6 above)
Comment (JulissaDantes) [reply src/transformations/utils/build-super-calls-for-chain.ts]: On the new commit the modifiers will never not exists as the intial state is an empty array

File: src/transformations/transform-constructor.ts [hunk 7]
```diff
@@ -96,29 +96,51 @@ export function* transformConstructor(
     if (constructorNode) {
       const { start: bodyStart } = getNodeBounds(constructorNode.body!);
       const argNames = constructorNode.parameters.parameters.map(p => p.name);
+      const parents = contractNode.linearizedBaseContracts;
+      const modifiers = [];
```
[the remark this answers is not in the document]
Comment (JulissaDantes) [reply src/transformations/transform-constructor.ts:119]: O the latest commit if the constructor doesnt have modifiers it will not run the loop

**Review by frangio — APPROVED** [review]
Thank you!

**Pull request merged**